nautobot/extras/models/groups.py

```diff
@@ -0,0 +1,283 @@
+"""Dynamic Groups Models."""
+
+import logging
+
+from django import forms
+from django.contrib.contenttypes.models import ContentType
+from django.core.exceptions import ValidationError
+from django.core.serializers.json import DjangoJSONEncoder
+from django.db import models
+from django.urls import reverse
+
+from nautobot.core.fields import AutoSlugField
+from nautobot.core.models.generics import OrganizationalModel
+from nautobot.extras.groups import dynamicgroup_map_factory
+from nautobot.extras.querysets import DynamicGroupQuerySet
+from nautobot.extras.utils import extras_features
+
+
+logger = logging.getLogger(__name__)
+
+
+@extras_features(
+    "custom_fields",
+    "custom_links",
+    "custom_validators",
+    "export_templates",
+    "graphql",
+    "relationships",
+    "webhooks",
+)
+class DynamicGroup(OrganizationalModel):
+    """Dynamic Group Model."""
+
+    name = models.CharField(max_length=100, unique=True, help_text="Dynamic Group name")
+    slug = AutoSlugField(max_length=100, unique=True, help_text="Unique slug", populate_from="name")
+    description = models.CharField(max_length=200, blank=True)
+    content_type = models.ForeignKey(
+        to=ContentType,
+        on_delete=models.CASCADE,
+        verbose_name="Object Type",
+        help_text="The type of object for this Dynamic Group.",
+    )
+    # TODO(jathan): Set editable=False that this field doesn't show up in forms.
+    # It's a complex field only modified by the DynamicGroupForm internals at
+    # this time. I am not happy with this pattern right now but due to the
+    # dynamism of the form by merging in the `FooFilterForm` fields, there's not
+    # currently an easy way to move this construction logic to the model.
+    filter = models.JSONField(
+        encoder=DjangoJSONEncoder,
+        editable=False,
+        default=dict,
+        blank=True,
+        null=True,
+        help_text="A JSON-encoded dictionary of filter parameters for group membership",
+    )
+
+    objects = DynamicGroupQuerySet.as_manager()
+
+    clone_fields = ["content_type", "filter"]
+
+    class Meta:
+        ordering = ["content_type", "name"]
+
+    def __str__(self):
+        return self.name
+
+    def get_queryset(self, **kwargs):
+        """
+        Define custom queryset for group model.
+
+        Any `kwargs` are passed along to `DynamicGroupMap.get_queryset()`.
+        """
+
+        model = self.content_type.model_class()
+
+        if not self.filter:
+            return model.objects.none()
+
+        qs = self.map.get_queryset(self.filter, **kwargs)
```

Comment:
Is this expensive? Should we do any sort of caching?

Reply:
Yeah it's probably expensive. I'll think about caching when I go to rip out the `BaseDynamicGroupMap` stuff in the next iteration. So I'l leave this comment unresolved.